File: 2022/12/python/main.py
# --- Constants --
INPUT_FILE = "../input/input.txt"

# --- Map class --
class Map():
    def __init__(self, data):
        self.grid = []
        for line in data.split("\n"):
            self.grid.append([ ord(c) for c in line ])

        self.height = len(self.grid)
        self.width = len(self.grid[0])
        s = ord("S")
        e = ord("E")
        for y in range(len(self.grid)):
            for x in range(len(self.grid[y])):
                if self.grid[y][x] == s:
                    self.start = (x, y)
                    self.grid[y][x] = ord("a")
                if self.grid[y][x] == e:
                    self.end = (x, y)
                    self.grid[y][x] = ord("z")

    # --- Find the shortest path ---
    def find_path(self):
        h = self.height
        w = self.width
        x = self.start[0]
        y = self.start[1]
        a = []
        for _ in range(h): a.append([0] * w);
        a[y][x] = 1
        queue = [ self.start ]
        while len(queue) > 0:
            p = queue.pop(0)
            x = p[0]
            y = p[1]
            if x > 0 and a[y][x - 1] == 0 and self.grid[y][x - 1] <= self.grid[y][x] + 1:
                a[y][x - 1] = a[y][x] + 1
                queue.append((x - 1, y))
            if y > 0 and a[y - 1][x] == 0 and self.grid[y - 1][x] <= self.grid[y][x] + 1:
                a[y - 1][x] = a[y][x] + 1
                queue.append((x, y - 1))
            if x < w - 1 and a[y][x + 1] == 0 and self.grid[y][x + 1] <= self.grid[y][x] + 1:
                a[y][x + 1] = a[y][x] + 1
                queue.append((x + 1, y))
            if y < h - 1 and a[y + 1][x] == 0 and self.grid[y + 1][x] <= self.grid[y][x] + 1:
                a[y + 1][x] = a[y][x] + 1
                queue.append((x, y + 1))

        x = self.end[0]
        y = self.end[1]
        c = a[y][x]
        queue.clear()
        while c > 1:
            if   x > 0     and a[y][x - 1] == c: x -= 1
            elif y > 0     and a[y - 1][x] == c: y -= 1
            elif x < w - 1 and a[y][x + 1] == c: x += 1
            elif y < h - 1 and a[y + 1][x] == c: y += 1
            queue.append((x, y))
            c -= 1

        return queue

    # --- Find shortest posible path ---
    def find_shortest_posible_path(self):
        h = self.height
        w = self.width
        m = 2**31 - 1
        for y in range(h):
            for x in range(w):
                if self.grid[y][x] == ord("a"):
                    self.start = (x, y)
                    steps = len(self.find_path())
                    if steps > 0: m = min(steps, m)
        return m


# --- Read and parse the input file --
file = open(INPUT_FILE, "r")
text = file.read().strip()
_map = Map(text)

# --- Puzzle 1 ---
count = len(_map.find_path())
print(f"1. Fewest steps: {count:,d}")

count = _map.find_shortest_posible_path()
print(f"2. Shortest path: {count:,d}")

File: 2022/12/c/main.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <queue.h>
#include <file_reader.h>

// --- Constants ---
#define INPUT_FILE "../input/input.txt"

// --- Structs ---
struct node { int x; int y; TAILQ_ENTRY(node) nodes; };
TAILQ_HEAD(head_s, node) head;
READ_BY_LINE_HEAD(file_t)

// --- Macros ---
#define map(y, x) map[(y) * w + (x)]

// Adds a new point into the queue
void queue_add(struct head_s* head, int x, int y) {
	struct node* e = malloc(sizeof(struct node));
	if (e == NULL) {
		fprintf(stderr, "Malloc failed");
		exit(EXIT_FAILURE);
	}
	e->x = x; e->y = y;
	TAILQ_INSERT_TAIL(head, e, nodes);
	e = NULL;
}

// Find path between two points
int find_path(const char* map, int w, int h, int start, int end) {
	int a[h][w];
	int x, y;
	for (y = 0; y < h; y++) for (x = 0; x < w; x++) a[y][x] = 0;

	x = start % w; y = start / w;
	a[y][x] = 1;

	TAILQ_INIT(&head);
	queue_add(&head, x, y);
	while(!TAILQ_EMPTY(&head)) {
		struct node* e = TAILQ_FIRST(&head);
		TAILQ_REMOVE(&head, e, nodes);
		x = e->x; y = e->y;
		free(e);
		e = NULL;

		if (x > 0     && a[y][x - 1] == 0 && map(y, x - 1) <= map(y, x) + 1) { a[y][x - 1] = a[y][x] + 1; queue_add(&head, x - 1, y); }
		if (y > 0     && a[y - 1][x] == 0 && map(y - 1, x) <= map(y, x) + 1) { a[y - 1][x] = a[y][x] + 1; queue_add(&head, x, y - 1); }
		if (x < w - 1 && a[y][x + 1] == 0 && map(y, x + 1) <= map(y, x) + 1) { a[y][x + 1] = a[y][x] + 1; queue_add(&head, x + 1, y); }
		if (y < h - 1 && a[y + 1][x] == 0 && map(y + 1, x) <= map(y, x) + 1) { a[y + 1][x] = a[y][x] + 1; queue_add(&head, x, y + 1); }
	}

	x = end % w; y = end / w;
	return a[y][x] - 1;
}

// Find the shortest path to the end point
int find_shortest_path(const char* map, int w, int h, int end) {
	int min = INT_MAX;
	int x, y;
	for (y = 0; y < h; y++) {
		for (x = 0; x < w; x++) {
			if (map(y, x) == 'a') {
				int start = y * w + x;
				int count = find_path(map, w, h, start, end);
				if (count > 0 && count < min) min = count;
			}
		}
	}
	return min;
}

// --- Main function ---
int main(void) {
	// --- Variables ---
	READ_BY_LINE_INIT(file, file_t, INPUT_FILE)
	char* map;
	int width = file.max - 2;
	int height = file.rows;

	// --- Read and parse the input file ---
	READ_BY_LINE_FULL(file, map)

	// --- Search start and end points ---
	char* ptr = strchr(map, 'S');
	if (ptr == NULL) {
		fprintf(stderr, "Start point not found!\n");
		exit(EXIT_FAILURE);
	}
	*ptr = 'a';
	int start = ptr - map;

	ptr = strchr(map, 'E');
	if (ptr == NULL) {
		fprintf(stderr, "End point not found!\n");
		exit(EXIT_FAILURE);
	}
	*ptr = 'z';
	int end = ptr - map;

	// --- Puzzle 1 ---
	int count = find_path(map, width, height, start, end);
	printf("1. Fewest steps: %d\n", count);

	// --- Puzzle 2 ---
	int min = find_shortest_path(map, width, height, end);
	printf("2. Shortest path: %d\n", min);

	// --- Free resources ---
	free(map); map = NULL;

	return EXIT_SUCCESS;
}
